lms/djangoapps/courseware/model_data.py

```diff
@@ -240,206 +821,133 @@ def _fields_to_cache(self, descriptors):
             # user we were constructed for.
             assert key.user_id == self.user.id
 
-        return self.cache.get(self._cache_key_from_kvs_key(key))
+        if key.scope not in self.cache:
+            raise KeyError(key.field_name)
 
-    def find_or_create(self, key):
-        '''
-        Find a model data object in this cache, or create it if it doesn't
-        exist
-        '''
-        field_object = self.find(key)
-
-        if field_object is not None:
-            return field_object
-
-        if key.scope == Scope.user_state:
-            field_object, __ = StudentModule.objects.get_or_create(
-                course_id=self.course_id,
-                student_id=key.user_id,
-                module_state_key=key.block_scope_id,
-                defaults={
-                    'state': json.dumps({}),
-                    'module_type': key.block_scope_id.block_type,
-                },
-            )
-        elif key.scope == Scope.user_state_summary:
-            field_object, __ = XModuleUserStateSummaryField.objects.get_or_create(
-                field_name=key.field_name,
-                usage_id=key.block_scope_id
-            )
-        elif key.scope == Scope.preferences:
-            field_object, __ = XModuleStudentPrefsField.objects.get_or_create(
-                field_name=key.field_name,
-                module_type=BlockTypeKeyV1(key.block_family, key.block_scope_id),
-                student_id=key.user_id,
-            )
-        elif key.scope == Scope.user_info:
-            field_object, __ = XModuleStudentInfoField.objects.get_or_create(
-                field_name=key.field_name,
-                student_id=key.user_id,
-            )
+        return self.cache[key.scope].get(key)
 
-        cache_key = self._cache_key_from_kvs_key(key)
-        self.cache[cache_key] = field_object
-        return field_object
+    @contract(kv_dict="dict(DjangoKeyValueStore_Key: *)")
+    def set_many(self, kv_dict):
+        """
+        Set all of the fields specified by the keys of `kv_dict` to the values
+        in that dict.
 
+        Arguments:
+            kv_dict (dict): dict mapping from `DjangoKeyValueStore.Key`s to field values
+        Raises: DatabaseError if any fields fail to save
+        """
 
-class DjangoKeyValueStore(KeyValueStore):
-    """
-    This KeyValueStore will read and write data in the following scopes to django models
-        Scope.user_state_summary
-        Scope.user_state
-        Scope.preferences
-        Scope.user_info
+        saved_fields = []
+        by_scope = defaultdict(dict)
+        for key, value in kv_dict.iteritems():
 
-    Access to any other scopes will raise an InvalidScopeError
+            if key.scope.user == UserScope.ONE and not self.user.is_anonymous():
+                # If we're getting user data, we expect that the key matches the
+                # user we were constructed for.
+                assert key.user_id == self.user.id
```

Comment:
This will abort the method entirely, and thus will make no changes to the stored data, which I think is the correct behavior. I don't have any tests that cover it, in part because I don't think that the assertion error is actually part of the published interface, as much as it's part of the contract of how the class should be used.